ui/src/shared/apis/queryCache.ts

```diff
@@ -0,0 +1,164 @@
+      if (this.cache[id].isCustomTime) {
+        continue
+      }
+      if (now - this.cache[id].dateSet > TIME_INVALIDATION) {
+        this.resetCacheByID(id)
+      }
+    }
+  }
+
+  getFromCache = (id: string, variableID: string): RunQueryResult | null => {
+    // no existing query match
+    if (!this.cache[id]) {
+      return null
+    }
+    // query match with no existing variable match
+    if (this.cache[id].variables !== variableID) {
+      this.resetCacheByID(id)
+      return null
+    }
+    // query & variable match with an expired result
+    if (Date.now() - this.cache[id].dateSet > TIME_INVALIDATION) {
+      this.resetCacheByID(id)
+      return null
+    }
+    return this.cache[id].values
+  }
+
+  resetCacheByID = (id: string): void => {
+    if (!this.cache[id]) {
+      return
+    }
+    delete this.cache[id]
+  }
+
+  resetCache = (): void => {
+    this.cache = {}
+  }
+
+  setCacheByID = (
+    queryID: string,
+    variableID: string,
+    values: RunQueryResult,
+    isCustomTime: boolean = false
+  ): void => {
+    this.cache[queryID] = {
+      dateSet: Date.now(),
+      isCustomTime,
+      values,
+      variables: variableID,
+    }
+  }
+
+  startWatchDog = () => {
+    setInterval(() => {
+      this._cleanExpiredQueries()
+    }, TIME_INVALIDATION / 2)
+
+    this._cleanExpiredQueries()
+  }
+}
+
+const queryCache = new QueryCache()
+// Set an interval to check for expired data to invalidate
+queryCache.startWatchDog()
+
+export const resetQueryCache = (): void => {
+  queryCache.resetCache()
+}
+
+export const resetQueryCacheByQuery = (query: string): void => {
+  const queryID = `${hashCode(query)}`
+  queryCache.resetCacheByID(queryID)
+}
+
+export const getRunQueryResults = (
+  orgID: string,
+  query: string,
+  state: AppState,
+  abortController?: AbortController
+): CancelBox<RunQueryResult> => {
+  const usedVars = filterUnusedVarsBasedOnQuery(getAllVariables(state), [query])
+  const variables = sortBy(usedVars, ['name'])
+  const simplifiedVariables = variables.map(v => asSimplyKeyValueVariables(v))
+  const stringifiedVars = JSON.stringify(simplifiedVariables)
+  // create the queryID based on the query & vars
+  const queryID = `${hashCode(query)}`
+  const variableID = `${hashCode(stringifiedVars)}`
+
+  const cacheResults: RunQueryResult | null = queryCache.getFromCache(
+    queryID,
+    variableID
+  )
+
+  // check the cache based on text & vars
+  if (cacheResults) {
+    const controller = abortController || new AbortController()
+    return {
+      promise: new Promise(resolve => resolve(cacheResults)),
+      cancel: () => controller.abort(),
+    }
+  }
+  const variableAssignments = variables
+    .map(v => asAssignment(v))
+    .filter(v => !!v)
+  // otherwise query & set results
+  const extern = buildVarsOption(variableAssignments)
+  const results = runQuery(orgID, query, extern, abortController)
+  results.promise = results.promise.then(res => {
```

Comment:
does this need to conform to the same return type as other query methods? it doesn't look like `results` is ever used, just the promise from it. why not just return the promise?

Reply:
I think I may be misunderstanding you, but we are returning the `results` here since it conforms specifically to the type that is expected to be returned in the function (a `CancelBox<RunQueryResult>`) which is an object with a `promise` and a `cancel` property.

Reply:
yeah, i figured that's why it's returned like this, so there isn't much to do.

what i was saying that i couldn't find a single place where code acted like this:

```js
const results = getRunQueryResults(orgID, myQuery, appState)

// uses results directly
if (results.foo) {
  foo()
}
```
it was always like this:

```ts
const results = getRunQueryResults(orgID, myQuery, appState)

// only use results.promise
results.promise.then(response => doSomething(response))
// or
// still only using results.promise
Promise.all(results, r => r.promise)
```